bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java

```diff
@@ -0,0 +1,508 @@
+            return;
+        }
+
+        // needed to get ThreadGroups working in JDK17 (openhab4). Otherwise, Python threads are not working properly.
+        // - if ThreadGroups.isDaemon == true and last python thread is done
+        // ==> ThreadGroup is closed and new python threads can't started anymore
+        // - in jdk21 is handled differently
+        // ==> ThreadGroups does not use the daemon flag anymore and ThreadGroups are not closable anymore
+        // - can be removed in jdk21 only environments
+        Thread.currentThread().getThreadGroup().setDaemon(false);
+
+        logger.debug("Initializing GraalPython script engine...");
+
+        ScriptContext ctx = getScriptContext();
+
+        // these are added post-construction, so we need to fetch them late
+        String engineIdentifier = (String) ctx.getAttribute(CONTEXT_KEY_ENGINE_IDENTIFIER);
+        if (engineIdentifier == null) {
+            throw new IllegalStateException("Failed to retrieve engine identifier from engine bindings");
+        }
+
+        ScriptExtensionAccessor scriptExtensionAccessor = (ScriptExtensionAccessor) ctx
+                .getAttribute(CONTEXT_KEY_EXTENSION_ACCESSOR);
+        if (scriptExtensionAccessor == null) {
+            throw new IllegalStateException("Failed to retrieve script extension accessor from engine bindings");
+        }
+
+        Consumer<String> scriptDependencyListener = (Consumer<String>) ctx
+                .getAttribute(CONTEXT_KEY_DEPENDENCY_LISTENER);
+        if (scriptDependencyListener == null) {
+            logger.warn(
+                    "Failed to retrieve script script dependency listener from engine bindings. Script dependency tracking will be disabled.");
+        }
+        this.scriptDependencyListener = scriptDependencyListener;
+
+        if (pythonScriptEngineConfiguration.isScopeEnabled()) {
+            // Wrap the "import" function to also allow loading modules from the ScriptExtensionModuleProvider
+            BiFunction<String, List<String>, Object> wrapImportFn = (name, fromlist) -> scriptExtensionModuleProvider
+                    .locatorFor(delegate.getPolyglotContext(), engineIdentifier, scriptExtensionAccessor)
+                    .locateModule(name, fromlist);
+            delegate.getBindings(ScriptContext.ENGINE_SCOPE).put(ScriptExtensionModuleProvider.IMPORT_PROXY_NAME,
+                    wrapImportFn);
+            try {
+                String wrapperContent = new String(
+                        Files.readAllBytes(PythonScriptEngineFactory.PYTHON_WRAPPER_FILE_PATH));
+                delegate.getPolyglotContext().eval(Source.newBuilder(GraalPythonScriptEngine.LANGUAGE_ID,
+                        wrapperContent, PythonScriptEngineFactory.PYTHON_WRAPPER_FILE_PATH.toString()).build());
+
+                // inject scope, Registry and logger
+                if (!pythonScriptEngineConfiguration.isInjection(PythonScriptEngineConfiguration.INJECTION_DISABLED)
```

Comment:
I really like that one can control where to automatically inject the library, that makes much sense. I will probably add this to JS Scripting as well.